src/lib.rs

```diff
@@ -336,18 +340,66 @@ impl Config {
         assert!(output.ends_with(".a"));
         let lib_name = &output[3..output.len() - 2];
         let dst = self.get_out_dir();
+        let msvc = self.get_target().contains("msvc");
+        let compiler = self.get_compiler();
 
         let mut objects = Vec::new();
         let mut src_dst = Vec::new();
+        let mut dependencies = HashMap::new();
         for file in self.files.iter() {
             let obj = dst.join(file).with_extension("o");
             let obj = if !obj.starts_with(&dst) {
                 dst.join(obj.file_name().unwrap())
             } else {
                 obj
             };
+
+            let mut input_mtime = get_mtime(&file);
+            let output_mtime = get_mtime(&obj);
+            // If input file is older than output object, check its dependencies
+            if input_mtime < output_mtime {
+                let mut cmd = compiler.to_command();
+                let files: Vec<PathBuf>;
+                cmd.arg(&file);
+                if msvc {
+                    cmd.arg("/showIncludes").arg("/E");
+                    println!("running: {:?}", cmd);
+                    files = cmd.output()
+                        .map(|o| String::from_utf8(o.stderr))
+                        .unwrap().unwrap()
+                        .replace("Note: including file:", "")
+                        .lines().skip(1) // skip input file
+                        .map(|s| PathBuf::from(s.trim()))
+                        .collect();
+                } else {
+                    cmd.arg("-M").arg("-MG");
+                    println!("running: {:?}", cmd);
+                    files = cmd.output()
+                        .map(|o| String::from_utf8(o.stdout))
+                        .unwrap().unwrap()
+                        .replace("\\\n", "") // remove line continuations
+                        .replace("\\ ", "\0") // handle spaces in filenames
+                        .split_whitespace().skip(2) // skip input/output files
```

Comment:
Unfortunately it's often multiple files per line, so it has to be split on non-escaped white space.